Lib/test/test_dis.py

```diff
@@ -1381,6 +1382,32 @@ def test__find_store_names(self):
                 res = tuple(dis._find_store_names(code))
                 self.assertEqual(res, expected)
 
+    def test_findlabels(self):
+        def zero_labels(x):
+            return 0
+        def one_label(x):
+            if x > 0:     # This will jump to "return 2" if not x > 0
+                return 1
+            else:
+                return 2
+        def two_labels(x):
+            if x > 0:
+                return 1
+            elif x > 10:
+                return 10
+            else:
+                return 2
+
+        cases = [
+            (zero_labels, []),
+            (one_label, [12]),
+            (two_labels, [12, 24]),
```

Comment:
where do the numbers come from?

Reply:
When we use `dis` on, let's say, `one_label`, here what we will see:

```
  2           0 LOAD_FAST                0 (x)
              2 LOAD_CONST               1 (0)
              4 COMPARE_OP               4 (>)
              6 POP_JUMP_IF_FALSE        6 (to 12)

  3           8 LOAD_CONST               2 (1)
             10 RETURN_VALUE

  5     >>   12 LOAD_CONST               3 (2)
             14 RETURN_VALUE
```

The source is `>>   12 LOAD_CONST` opcode (which we jump to from `6 POP_JUMP_IF_FALSE        6 (to 12)`).

Here's the `dis` for `two_labels`:

```
  2           0 LOAD_FAST                0 (x)
              2 LOAD_CONST               1 (0)
              4 COMPARE_OP               4 (>)
              6 POP_JUMP_IF_FALSE        6 (to 12)

  3           8 LOAD_CONST               2 (1)
             10 RETURN_VALUE

  4     >>   12 LOAD_FAST                0 (x)
             14 LOAD_CONST               3 (10)
             16 COMPARE_OP               4 (>)
             18 POP_JUMP_IF_FALSE       12 (to 24)

  5          20 LOAD_CONST               3 (10)
             22 RETURN_VALUE

  7     >>   24 LOAD_CONST               4 (2)
             26 RETURN_VALUE
```

Reply:
This seems brittle, in the sense that many changes to bytecode generation could break this test and require updating it.

If possible, it would be good to find a way to make this test more robust (perhaps using other `dis` module functionality to find "RETURN_VAULE" op codes followed by "LOAD_FAST" or "LOAD_CONST"?).

Also, it seems that a comment explaining the details here would be helpful.

Reply:
> if possible, it would be good to find a way to make this test more robust (perhaps using other dis module functionality to find "RETURN_VAULE" op codes followed by "LOAD_FAST" or "LOAD_CONST"?).

I like the idea of making this test more robust, but seems like the proposed solution won't work 🤔 

Here's how `findlabels` is implemented:

```python
def findlabels(code):
    labels = []
    for offset, op, arg in _unpack_opargs(code):
        if arg is not None:
            if op in hasjrel:
                label = offset + 2 + arg*2
            elif op in hasjabs:
                label = arg*2
            else:
                continue
            if label not in labels:
                labels.append(label)
    return labels
```

Where:
- `hasjrel` are `opcode.hasjrel` (which has entries: `FOR_ITER`, `JUMP_FORWARD`, `SEND`)
- `hasjabs`: `JUMP_IF_FALSE_OR_POP`, `JUMP_IF_TRUE_OR_POP`, and many more

I don't think that we can safely find some other algorithm to describe `findlabels()`.

> This seems brittle, in the sense that many changes to bytecode generation could break this test and require updating it.

But, since `dis()` itself prints the same info in its outputs (it even uses [`findlabels()` as a helper](https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/dis.py#L403)):
1. https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/test/test_dis.py#L541
2. https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/test/test_dis.py#L547
3. https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/test/test_dis.py#L347
4. https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/test/test_dis.py#L329
5. https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/test/test_dis.py#L143
6. https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/test/test_dis.py#L146

I guess in case of a major refactoring, all these test cases will have to be changed. And in my point of view, this is a good thing to have more tests like these, since some code introspection / coverage / etc tools might rely on this behavior. And the change's author will double check if the change is correct.

One last thing, as I said, `dis` uses `findlabels` inside, maybe we can completely skip testing it? 🤔 
In my opinion, no - we can't do that. Because `dis()` duplicates the same logic here: https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/dis.py#L430-L435

Leaving `findlabels()` with a single responsibility: https://github.com/python/cpython/blob/243c31667cc15a9a338330ad9b2a29b1cd1c76ec/Lib/dis.py#L413

So, it needs to be tested explicitly.

> Also, it seems that a comment explaining the details here would be helpful.

I totally agree that I need to add a better explanation to what is going on in this test, since there are already two people asking 🙂

Reply:
I am also going to cover `if op in hasjrel:` branch in `findlabels`, because right now I only test `elif op in hasjabs:` branch.

Reply:
Looking more broadly at the existing tests for this module...

Have you considered using the existing `jumpy()` function for testing `findlabels()`? Possibly using the detailed information in `expected_opinfo_jumpy`?

Reply:
Or maybe we can just make sure that all `>> X` in `dis()` match the results found by `findlabels()` 🤔

Reply:
No, but, this seems like a great solution! 👍